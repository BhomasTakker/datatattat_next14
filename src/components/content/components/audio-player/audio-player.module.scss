@use "../../../../scss/mixins/text/" as textMixins;
@use "../../../../scss/abstracts/colours/" as *;
@use "../../../../scss/abstracts/units/" as *;

.audioPlayer {
	--bar-bg: #dcdcdc;
	--seek-before-handle: #696969;
	--handle: #696969;
	--handle-selcted: #696969;

	.progressBarContainer {
		padding: $rem_3 $rem_5;
		background-color: $whiterSmoke;
		border: 4px solid $whiteSmoke;
		border-radius: $rem_1;
	}
	.controls {
		display: flex;
		flex-direction: column;
		align-items: center;

		width: 100%;
		background-color: $whiteSmoke;

		.playerControls {
			display: flex;
			align-items: center;
			// width: 100%;
			padding: $rem_1 $rem_2;
		}

		.controlsSecondary {
			display: flex;
			align-items: center;
			justify-content: space-between;
			width: 100%;
			// 	padding: $rem_1 $rem_2;
			.speakerContainer {
				display: flex;
				align-items: center;
			}
		}

		.volumeButton {
			// padding-right: 1rem;
		}

		.volumeBg {
			padding-left: 0rem;
			cursor: unset;
		}
		.volumeBg:hover {
			background-color: $whiteSmoke;
		}
	}

	.timeBg {
		cursor: unset;
		&:hover {
			background-color: $whiteSmoke;
		}

		// .time {
		// 	@include textMixins.maxLines(1);
		// }
	}

	input[type="range"] {
		appearance: none;
		background: var(--bar-bg);
		border-radius: $rem_2;
		position: relative;
		width: 100%;
		height: $rem_1;
		outline: none;
	}

	// safari
	input[type="range"]::-webkit-slider-runnable-track {
		background: var(--bar-bg);
		border-radius: $rem_2;
		position: relative;
		width: 100%;
		height: $rem_1;
		outline: none;
	}

	// firefox
	input[type="range"]::-moz-range-track {
		background: var(--bar-bg);
		border-radius: $rem_2;
		position: relative;
		width: 100%;
		height: $rem_1;
		outline: none;
	}

	input[type="range"]::-moz-focus-outer {
		border: 0;
	}

	// progress bar chrome & Safari
	input[type="range"]::before {
		content: "";
		height: $rem_1;
		background: var(--seek-before-handle);
		border-top-left-radius: $rem_2;
		border-bottom-left-radius: $rem_2;
		position: absolute;
		top: 0;
		left: 0;
		z-index: 2;
		cursor: pointer;
	}

	.progressBar {
		--playback-width: 0;
	}

	.progressBar::before {
		width: var(--playback-width);
	}

	.volume {
		--volume-width: 50%;
	}

	.volume::before {
		width: var(--volume-width);
	}

	input[type="range"]::-moz-range-progress {
		background: var(--seek-before-handle);
		border-top-left-radius: $rem_2;
		border-bottom-left-radius: $rem_2;
		position: absolute;
		height: $rem_1;
	}

	// handle
	input[type="range"]::-webkit-slider-thumb {
		-webkit-appearance: none;
		width: $rem_3;
		height: $rem_3;
		background: var(--handle);
		border-radius: 50%;
		border: none;
		cursor: pointer;
		position: relative;
		margin: -3px 0 0 0;
		z-index: 3;
		box-sizing: border-box;
	}

	input[type="range"]:active::-webkit-slider-thumb {
		transform: scale(1.2);
		background: var(--handle-selcted);
	}

	input[type="range"]::-moz-range-thumb {
		width: $rem_3;
		height: $rem_3;
		background: var(--handle);
		border-radius: 50%;
		border: transparent;
		cursor: pointer;
		position: relative;
		z-index: 3;
		box-sizing: border-box;
	}

	input[type="range"]:active::-moz-range-thumb {
		transform: scale(1.2);
		background: var(--handle-selcted);
	}

	@media screen and (min-width: $screenWidthSm) {
		.controls {
			flex-direction: row;
		}
	}
}
