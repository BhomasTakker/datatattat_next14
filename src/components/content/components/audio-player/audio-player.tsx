"use client";

// https://www.youtube.com/watch?v=sqpg1qzJCGQ&list=PLrz61zkUHJJGLD5qlJSYfbm0Gh1Ri1Rt0&index=1
import styles from "./audio-player.module.scss";

import ReactH5AudioPlayer, { RHAP_UI } from "react-h5-audio-player";
import "react-h5-audio-player/src/styles.scss";
import "video.js/dist/video-js.css";
import "videojs-youtube";

import H5AudioPlayer from "react-h5-audio-player";
import { VideoDisplayComponent } from "../display-player/display-player";

///////////////////////////////////////////
// Take in any audio player controls etc
// I guess if playlist etc
// but for now just a simple player
///////////////////////////////////////////
type AudioPlayerProps = {
	src: string;
	variant?: "default" | "radio" | string;
};

type RendeProps = {
	src: string;
	variant?: "default" | "radio" | string;
};

// A bit of a hack - we need a better solution
const renderPlayer = ({ src, variant }: RendeProps) => {
	switch (true) {
		case src.endsWith(".m3u8"):
			return (
				<VideoDisplayComponent
					key={src}
					options={{ src, poster: "", audioOnlyMode: true }}
				/>
			);
		case variant === "radio":
			return (
				<ReactH5AudioPlayer
					src={src}
					showJumpControls={false}
					layout="horizontal-reverse"
					customControlsSection={[
						RHAP_UI.MAIN_CONTROLS,
						RHAP_UI.VOLUME_CONTROLS,
					]}
					autoPlayAfterSrcChange={true}
					// showDownloadProgress={false}
					showFilledVolume={true}
					// defaultDuration={null}
				/>
			);
		default:
			return (
				<ReactH5AudioPlayer
					src={src}
					layout="horizontal-reverse"
					autoPlayAfterSrcChange={false}
					// showDownloadProgress={false}
					showFilledVolume={true}
					defaultDuration={null}
				/>
			);
	}
};

export const AudioPlayer = ({ src, variant }: AudioPlayerProps) => {
	return (
		<div className={styles.audioPlayer}>{renderPlayer({ src, variant })}</div>
	);
};
