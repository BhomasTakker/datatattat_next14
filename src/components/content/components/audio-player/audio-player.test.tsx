import React, { createRef } from "react";
import { render, fireEvent, act } from "@testing-library/react";
import { AudioPlayer } from "./audio-player";

const mockSrc = "test-audio.mp3";

function setup() {
	const audioPlayerRef = createRef<HTMLAudioElement>();
	const utils = render(
		<AudioPlayer src={mockSrc} audioPlayerRef={audioPlayerRef} />
	);
	const playPauseButton = utils.getAllByRole("button")[1];
	const back30Button = utils.getAllByRole("button")[0];
	const forward30Button = utils.getAllByRole("button")[2];
	const volumeButton = utils.getAllByRole("button")[3];
	const volumeSlider = utils.container.querySelector(
		'input[type="range"].' + "volume"
	);
	const progressBar = utils.container.querySelector(
		'input[type="range"].' + "progressBar"
	);
	return {
		...utils,
		audioPlayerRef,
		playPauseButton,
		back30Button,
		forward30Button,
		volumeButton,
		volumeSlider,
		progressBar,
	};
}

describe("AudioPlayer", () => {
	beforeEach(() => {
		jest.clearAllMocks();
	});

	it("renders audio element with correct src", () => {
		const { container } = setup();
		const audio = container.querySelector("audio");
		expect(audio).toBeInTheDocument();
		expect(audio?.getAttribute("src")).toBe(mockSrc);
	});

	it("toggles play/pause when playPauseButton is clicked", () => {
		const { playPauseButton, audioPlayerRef } = setup();
		const play = jest.fn();
		const pause = jest.fn();
		// @ts-ignore
		audioPlayerRef.current = {
			play,
			pause,
			paused: true,
			currentTime: 0,
			duration: 100,
			volume: 0.5,
			muted: false,
			addEventListener: jest.fn(),
			removeEventListener: jest.fn(),
		};
		// Play
		fireEvent.click(playPauseButton);
		expect(play).toHaveBeenCalled();
		// Pause
		fireEvent.click(playPauseButton);
		expect(pause).toHaveBeenCalled();
	});

	it("calls back30Handler and forward30Handler", () => {
		const { back30Button, forward30Button, progressBar, audioPlayerRef } =
			setup();
		// @ts-ignore
		audioPlayerRef.current = {
			currentTime: 50,
			duration: 100,
			volume: 0.5,
			muted: false,
			addEventListener: jest.fn(),
			removeEventListener: jest.fn(),
		};
		if (progressBar) {
			// @ts-ignore - Property 'value' does not exist on type 'Element'
			progressBar.value = "50";
			fireEvent.click(back30Button);
			// @ts-ignore - Property 'value' does not exist on type 'Element'
			expect(Number(progressBar.value)).toBe(20);
			// @ts-ignore - Property 'value' does not exist on type 'Element'
			progressBar.value = "50";
			fireEvent.click(forward30Button);
			// @ts-ignore - Property 'value' does not exist on type 'Element'
			expect(Number(progressBar.value)).toBe(80);
		}
	});

	it("toggles mute when volumeButton is clicked", () => {
		const { volumeButton, audioPlayerRef } = setup();
		// @ts-ignore
		audioPlayerRef.current = {
			muted: false,
			volume: 0.5,
			addEventListener: jest.fn(),
			removeEventListener: jest.fn(),
		};
		fireEvent.click(volumeButton);
		// @ts-ignore - current is possibly null
		expect(audioPlayerRef.current.muted).toBe(true);
		fireEvent.click(volumeButton);
		// @ts-ignore - current is possibly null
		expect(audioPlayerRef.current.muted).toBe(false);
	});

	it("changes volume when volume slider is changed", () => {
		const { volumeSlider, audioPlayerRef } = setup();
		// @ts-ignore
		audioPlayerRef.current = {
			volume: 0.5,
			muted: false,
			addEventListener: jest.fn(),
			removeEventListener: jest.fn(),
		};
		if (volumeSlider) {
			fireEvent.change(volumeSlider, { target: { value: "80" } });
			// @ts-ignore - current is possibly null
			expect(audioPlayerRef.current.volume).toBeCloseTo(0.8);
		}
	});

	it("updates currentTime and progress bar on range change", () => {
		const { progressBar, audioPlayerRef } = setup();
		// @ts-ignore
		audioPlayerRef.current = {
			currentTime: 0,
			duration: 100,
			volume: 0.5,
			muted: false,
			addEventListener: jest.fn(),
			removeEventListener: jest.fn(),
		};
		if (progressBar) {
			fireEvent.change(progressBar, { target: { value: "30" } });
			// @ts-ignore - current is possibly null
			expect(audioPlayerRef.current.currentTime).toBe(30);
		}
	});

	it("displays formatted time", () => {
		const { container, audioPlayerRef } = setup();
		// @ts-ignore - missing ....
		audioPlayerRef.current = {
			currentTime: 65,
			duration: 125,
			volume: 0.5,
			muted: false,
			addEventListener: jest.fn(),
			removeEventListener: jest.fn(),
		};
		// Not sure what copilot was trying to do here, but it was causing issues
		// act(() => {
		// 	fireEvent(audioPlayerRef.current as any, new Event("play"));
		// });
		const timeDisplay = container.querySelector("." + "time");
		expect(timeDisplay?.textContent).toMatch(/\d{2}:\d{2} \/ \d{2}:\d{2}/);
	});

	describe("Snaps", () => {
		it("matches snapshot", () => {
			const { asFragment } = setup();
			expect(asFragment()).toMatchSnapshot();
		});
	});
});
