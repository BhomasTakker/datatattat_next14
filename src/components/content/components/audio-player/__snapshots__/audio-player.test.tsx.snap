// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`AudioPlayer Snaps matches snapshot for .m3u8 src (VideoDisplayComponent) 1`] = `
<div
  class="audioPlayer"
>
  <div
    data-testid="video-display-component"
    options="[object Object]"
  >
    VideoDisplayComponent
    <span
      data-testid="video-src"
    >
      stream.m3u8
    </span>
    <span
      data-testid="audio-only"
    >
      true
    </span>
  </div>
</div>
`;

exports[`AudioPlayer Snaps matches snapshot for custom variant 1`] = `
<div
  class="audioPlayer"
>
  <div
    data-testid="h5-audio-player"
    layout="horizontal-reverse"
    src="audio.mp3"
  >
    H5AudioPlayer
    <span
      data-testid="src"
    >
      audio.mp3
    </span>
  </div>
</div>
`;

exports[`AudioPlayer Snaps matches snapshot for default audio player 1`] = `
<div
  class="audioPlayer"
>
  <div
    data-testid="h5-audio-player"
    layout="horizontal-reverse"
    src="audio.mp3"
  >
    H5AudioPlayer
    <span
      data-testid="src"
    >
      audio.mp3
    </span>
  </div>
</div>
`;

exports[`AudioPlayer Snaps matches snapshot for radio variant 1`] = `
<div
  class="audioPlayer"
>
  <div
    customcontrolssection="MAIN_CONTROLS,VOLUME_CONTROLS"
    data-testid="h5-audio-player"
    layout="horizontal-reverse"
    src="radio.mp3"
  >
    H5AudioPlayer
    <span
      data-testid="src"
    >
      radio.mp3
    </span>
  </div>
</div>
`;
